internal/util/basesubjectset.go

```diff
@@ -3,210 +3,230 @@ package util
 import (
 	"golang.org/x/exp/maps"
 
+	v1 "github.com/authzed/spicedb/pkg/proto/dispatch/v1"
 	"github.com/authzed/spicedb/pkg/tuple"
 )
 
-// Subject is a subject that can be placed into a BaseSubjectSet.
-type Subject interface {
+// Subject is a subject that can be placed into a BaseSubjectSet. It is defined in a generic
+// manner to allow implementations that wrap BaseSubjectSet to add their own additional bookkeeping
+// to the base implementation.
+type Subject[T any] interface {
 	// GetSubjectId returns the ID of the subject. For wildcards, this should be `*`.
 	GetSubjectId() string
 
-	// GetExcludedSubjectIds returns the list of subject IDs excluded. Should only have values
-	// for wildcards.
-	GetExcludedSubjectIds() []string
+	// GetConditionalExpression returns the caveat expression for this subject, if it is conditional.
+	GetConditionalExpression() *v1.CaveatExpression
```

Comment:
nit: why not just call this `GetCaveatExpression`?

Reply:
I don't mind changing it

Reply:
Changed